### PR TITLE
Moved path util code from jupyter_client and jupyter_server to core

### PR DESCRIPTION
**Pull request by MSeal**
Moved shared code from https://github.com/jupyter/jupyter_server/pull/77 and https://github.com/jupyter/jupyter_client/pull/469 as well as some other path related utils in jupyrer_server that seemed useful in a more generic scope. No code changes made on the functions  other than converting the tests from nose.assert pattern to pytest assert pattern.

Tested on windows & ubuntu independently

### CODE REVIEW

File: jupyter_core/paths.py [hunk 1]
```diff
@@ -187,3 +196,213 @@ def jupyter_config_path():
...
+        # Touch file pre-emptively to avoid editing permissions in open files in Windows
+        fd = os.open(fname, os.O_CREAT | os.O_WRONLY | os.O_TRUNC, 0o600)
+        os.close(fd)
+        open_flag = os.O_WRONLY | os.O_TRUNC
+        win32_restrict_file_to_user(fname)
+
+    with os.fdopen(os.open(fname, open_flag, 0o600), mode) as f:
+        if os.name != 'nt':
+            # Enforce that the file got the requested permissions before writing
+            assert '0600' == oct(stat.S_IMODE(os.stat(fname).st_mode)).replace('0o', '0')
```
Comment (vidartf) [inline jupyter_core/paths.py]: Do we really not care for this check if running in optimized mode (since it uses `assert` instead of `if`/`raise`)?

Comment (MSeal) [reply jupyter_core/paths.py]: It should probably be a raise and not an assert, though it's not critical to do this check. I'll make the change tomorrow.

Comment (vidartf) [reply jupyter_core/paths.py]: Sure, I was just linting a bit while waiting for someone else to have a look at it. The comment did say "enforce", which made it sound a little more critical 🤷‍♂

Comment (MSeal) [reply jupyter_core/paths.py]: Np, this is what code reviews are for. I've updated the assertion check.

**Comment by minrk** [comment]
Makes sense, thanks

**Pull request merged**